src/pg_query_summary.c

```diff
@@ -0,0 +1,761 @@
+				add_function(list_make1(makeString(stmt->newname)), CONTEXT_DDL, state);
+			}
+
+			return false; // Not handled by tree walker.
+		}
+
+		case T_PrepareStmt:
+			return pg_query_summary_walk_impl(castNode(PrepareStmt, node)->query, state);
+
+		case T_RawStmt:
+			return pg_query_summary_walk_impl(castNode(RawStmt, node)->stmt, state);
+
+		// The following are added to avoid needing subselect_items,
+		// like is used in Rust and Ruby.
+
+		case T_A_Expr: {
+			A_Expr* expr = castNode(A_Expr, node);
+			pg_query_summary_walk_impl(expr->lexpr, state);
+			pg_query_summary_walk_impl(expr->rexpr, state);
+
+			return false; // Not handled by tree walker.
+		}
+
+		case T_ColumnRef: {
+			if (!state->save_filter_columns)
+				break;
+
+			ColumnRef* ref = castNode(ColumnRef, node);
+
+			String* schema_name = NULL;
+			String* table_name = NULL;
+			String* column = NULL;
+
+			ListCell* lc = NULL;
+			foreach(lc, ref->fields) {
+				Node* field_node = lfirst(lc);
+
+				// If it's not a String, skip it.
+				if (field_node && field_node->type != T_String)
+					continue;
+
+				String* field = castNode(String, field_node);
+
+				if (!column) {
+					column = field;
+				} else if (!table_name) {
+					table_name = column;
+					column = field;
+				} else {
+					schema_name = table_name;
+					table_name = column;
+					column = field;
+				}
+			}
+
+			// If there's no clumn information, skip it.
+			if (!column)
+				break;
+
+			FilterColumn* fc = palloc(sizeof(FilterColumn));
+			fc->schema_name = schema_name ? schema_name->sval : NULL;
+			fc->table_name = table_name ? table_name->sval : NULL;
+			fc->column = column ? column->sval : NULL;
+
+			state->filter_columns = lappend(state->filter_columns, fc);
+
+			break;
+		}
+
+		// Things which we are only handling because raw_expression_tree_walker()
+		// doesn't handle them.
+
+		case T_CheckPointStmt:
+		case T_CreateSchemaStmt:
+		case T_GrantRoleStmt:
+		case T_TransactionStmt:
+		case T_VariableShowStmt:
+		case T_VariableSetStmt:
+			return false;
+
+		default:
+			break;
+	}
+
+	return raw_expression_tree_walker(node, pg_query_summary_walk_impl, (void*) state);
+}
+
+static char* range_var_parts_to_name(char* schemaname, char* relname) {
+	StringInfoData buf;
+	initStringInfo(&buf);
+	if (schemaname != NULL) {
+		appendStringInfoString(&buf, quote_identifier(schemaname));
+		appendStringInfoChar(&buf, '.');
+	}
+	appendStringInfoString(&buf, quote_identifier(relname));
+	return buf.data;
+}
+
+// Walk through all collected cte_names, and compares them to relname.
+//
+// Returns true if any are an exact match, false otherwise.
+static bool cte_names_include(WalkState* state, char* relname) {
+	ListCell* lc = NULL;
+	foreach(lc, state->cte_names) {
+		char *ctename = lfirst(lc);
+		if (strcmp(ctename, relname) == 0)
+			return true;
+	}
+	return false;
+}
+
+static void add_table(WalkState* state, ContextType context, char* schemaname, char* relname) {
```

Comment:
We should probably locate this close to `add_function`, since they behave similarly.

Reply:
Also, if this takes a `RangeVar` directly, we could simplify the calling code.